src/NotaryService/NotaryService.cs

```diff
@@ -0,0 +1,333 @@
+            for (int i = 0; i < payload.MainTransaction.Witnesses.Length; i++)
+            {
+                if (payload.MainTransaction.Witnesses[i].InvocationScript.Length == 0 || r.WitnessInfo[i].NSigsLeft == 0 && r.WitnessInfo[i].Typ != RequestType.Contract)
+                    continue;
+                switch (r.WitnessInfo[i].Typ)
+                {
+                    case RequestType.Contract:
+                        if (!r.MainTx.VerifyWitness(neoSystem.Settings, snapshot, r.MainTx.Signers[i].Account, payload.MainTransaction.Witnesses[i], r.MainTx.NetworkFee, out _))
+                            continue;
+                        r.MainTx.Witnesses[i].InvocationScript = payload.MainTransaction.Witnesses[i].InvocationScript;
+                        break;
+                    case RequestType.Signature:
+                        if (Crypto.VerifySignature(mainHash, payload.MainTransaction.Witnesses[i].InvocationScript.Skip(2).ToArray(), r.WitnessInfo[i].Pubs[0]))
+                        {
+                            r.MainTx.Witnesses[i] = payload.MainTransaction.Witnesses[i];
+                            r.WitnessInfo[i].NSigsLeft--;
+                        }
+                        break;
+                    case RequestType.MultiSignature:
+                        r.WitnessInfo[i].Sigs ??= new Dictionary<ECPoint, byte[]>();
+                        foreach (var pub in r.WitnessInfo[i].Pubs)
+                        {
+                            if (r.WitnessInfo[i].Sigs.ContainsKey(pub))
+                                continue;
+                            if (Crypto.VerifySignature(mainHash, payload.MainTransaction.Witnesses[i].InvocationScript.Skip(2).ToArray(), pub))
+                            {
+                                r.WitnessInfo[i].Sigs[pub] = payload.MainTransaction.Witnesses[i].InvocationScript;
+                                r.WitnessInfo[i].NSigsLeft--;
+                                if (r.WitnessInfo[i].NSigsLeft == 0)
+                                {
+                                    byte[] invScript = Array.Empty<byte>();
+                                    foreach (var t in r.WitnessInfo[i].Pubs)
+                                    {
+                                        if (r.WitnessInfo[i].Sigs.TryGetValue(t, out var sig))
+                                            invScript = invScript.Concat(sig).ToArray();
+                                    }
+                                    r.MainTx.Witnesses[i].InvocationScript = invScript;
+                                }
+                                break;
+                            }
+                        }
+                        break;
+                }
+                var currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+                if (r.IsMainCompleted() && r.MinNotValidBefore > currentHeight)
+                {
+                    Finalize(r.MainTx);
+                    r.IsSent = true;
```

Comment:
The same note about `isSent`, this field should be updated only after main tx persisting.